### ISSUE TITLE
AssertionError in ArrayTheory.merge()

### ISSUE THREAD

**Issue by PhilippWendler** [ultimate-pa/smtinterpol#14]
We have found an AssertionError when solving some queries with arrays:
```
java.lang.AssertionError
        at de.uni_freiburg.informatik.ultimate.smtinterpol.theory.cclosure.ArrayTheory.merge(ArrayTheory.java:951)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.theory.cclosure.ArrayTheory.buildWeakEq(ArrayTheory.java:1099)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.theory.cclosure.ArrayTheory.checkpoint(ArrayTheory.java:578)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.dpll.DPLLEngine.propagateInternal(DPLLEngine.java:246)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.dpll.DPLLEngine.solve(DPLLEngine.java:1026)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.dpll.DPLLEngine$AllSatIterator.hasNext(DPLLEngine.java:1635)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.ParseEnvironment.printResponse(ParseEnvironment.java:205)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.Parser$Action$.CUP$do_action(Parser.java:2181)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.Parser.do_action(Parser.java:853)
        at java_cup.runtime.LRParser.parse(Unknown Source)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.ParseEnvironment.parseStream(ParseEnvironment.java:129)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.ParseEnvironment.parseScript(ParseEnvironment.java:114)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.smtlib2.SMTLIB2Parser.run(SMTLIB2Parser.java:40)
        at de.uni_freiburg.informatik.ultimate.smtinterpol.Main.main(Main.java:182)
```

This happens with the latest version 2.1-235-g3b87807. An input file for reproducing it is here: https://gist.github.com/PhilippWendler/7bfce5387a4c22b938f3 (it happens with a few different queries, but the stacktrace is always identical, thus I assume it is the same problem).

**Comment by jhoenicke** [comment]
I have to look into this further.  As far as I see until now, the array indices are of the form `4*i+a` and `4*s+a` and are currently not equal.  When we want to add the inequality `4*i+a = 4*s+a`, it gets simplified to `i=s` and that equality is already set to true.  This breaks the assumption that the array theory knows the equalities on the index terms.  I have to think about whether we want to eagerly propagate these equalities (how?) or whether we can live without the assumption in the array theory.

Here is a simplified test that triggers the bug:

```
(set-logic QF_AUFLIRA)
(declare-fun |__ADDRESS_OF_main::a| () Int)
(declare-fun |main::i@2| () Int)
(declare-fun *signed_int@2 () (Array Int Int))
(declare-fun |main::s@2| () Int)
(declare-fun |main::tmp@2| () Int)
(declare-fun *signed_int@3 () (Array Int Int))
(declare-fun *signed_int@4 () (Array Int Int))
(assert (or (let ((.cse0 (+ |__ADDRESS_OF_main::a| (* 4 |main::s@2|)))) (and (= |main::tmp@2| (select *signed_int@2 .cse0)) (= *signed_int@3 (store *signed_int@2 .cse0 0)) (= *signed_int@4 (store *signed_int@3 (+ |__ADDRESS_OF_main::a| (* 4 |main::i@2|)) 0)))) (and (= |main::s@2| |main::i@2|) (= *signed_int@4 *signed_int@2))))
(check-sat)
```

**Comment by jhoenicke** [comment]
It turned out to be easier than I thought.  We just have to create the right equality atom and the code is already there.  Fixed by e5a0f5a951574abd702282b07cbd9fbd12f50889.

**Comment by PhilippWendler** [comment]
Thanks for the fast fix! It solves all instances of the AssertionError we have seen.
